### DISCUSSION TITLE
Generator: 数组变量作为函数参数的问题

### DISCUSSION THREAD

**Discussion by ma-h-m** [arcsysu/SYsU-lang#103] [Q&A]
对于下段代码：

```
int b[5][5];

int func(int a[5][5]){

    int sum=0;

    sum = b[2][3];
    sum = a[2][3];

    return sum;
}
```

对全局变量数组与函数参数访问方式分别如下：

```
  %0 = load i32, i32* getelementptr inbounds ([5 x [5 x i32]], [5 x [5 x i32]]* @b, i64 0, i64 2, i64 3), align 4
  store i32 %0, i32* %sum, align 4


  %1 = load [5 x i32]*, [5 x i32]** %a.addr, align 8
  %arrayidx = getelementptr inbounds [5 x i32], [5 x i32]* %1, i64 2
  %arrayidx1 = getelementptr inbounds [5 x i32], [5 x i32]* %arrayidx, i64 0, i64 3
  %2 = load i32, i32* %arrayidx1, align 4
  store i32 %2, i32* %sum, align 4
```

可以请助教讲解一下当数组作为参数时，访问数组的规则吗？

**Comment by zejia-lin** [comment]
高维数组的GEP操作较为复杂，可以将高维数组看作一维数组操作，也能够通过测试。

高维数组作为参数的时候，是一个“指针的数组”，即`[5 x i32]*`，相当于C语言的`int *a[5]`，所以首先需要`%arrayidx`来获取内层数组的地址。至于`%arrayidx1`和`%0`的GEP中为什么有0，可以参考[官方文档](https://llvm.org/docs/GetElementPtr.html)和[这篇博客](https://blog.yossarian.net/2020/09/19/LLVMs-getelementptr-by-example)。简单的来说，在C语言中“数组”可以看成指针，但LLVM的“数组”是一种类型，不是指针，`@b`和`%arrayidx`都是指向此“数组”的指针。也就是说对于foo[i]的访问，需要使用 (&foo)[0][i]的方式来完成，第一个参数自然是0。
